File: Python/astar.py
import time
from math import sqrt
from typing import List, Tuple
from sys import stdout
import numpy as np
import random
from termcolor import colored

# 1 for manhattan, 0 for euclidean
HEURISTIC = 0
DIAGONALS = True
TIME = True
PLOT = True

class World:
    def __init__(self, length: int, height: int, p_walls: float):
        self.length = length
        self.height = height
        self.p_walls = p_walls
        self.path_added = False

        self.grid = np.zeros((height, length), int)
        self.__generate_walls()

        if not DIAGONALS:
            # up, left, down, right
            self.delta = [[-1, 0, 1], 
                          [0, -1, 1], 
                          [1, 0, 1], 
                          [0, 1, 1]]  
        else:
            # up, left, down, right 
            # upleft, upright, downleft, downright
            self.delta = [[-1, 0, 1], 
                          [0, -1, 1], 
                          [1, 0, 1], 
                          [0, 1, 1],
                          [-1, -1, sqrt(2)],
                          [-1, 1, sqrt(2)],
                          [1, -1, sqrt(2)],
                          [1, 1, sqrt(2)]]

    def add_path(self, path: List[Tuple[int]]):
        i = 0
        for elem in path:
            if i == 0 or i == len(path) - 1:
                self.grid[elem[0]][elem[1]] = 3
            else:
                self.grid[elem[0]][elem[1]] = 2
            i += 1
        self.path_added = True

    def plot_grid(self):
        for rows in self.grid:
            for elem in rows:
                if elem == 1:
                    stdout.write(colored('█', 'red'))
                elif elem == 0:
                    stdout.write(' ')
                elif elem == 2:
                    stdout.write(colored('¤', 'blue'))
                elif elem == 3:
                    stdout.write(colored('x', 'green'))
                else:
                    stdout.write('o', 'yellow')
            print()

    def get_random_available_position(self) -> Tuple[int]:
        i = 0
        while 1:
            i += 1
            random_row = random.randint(0, self.height - 1)
            random_col = random.randint(0, self.length - 1)

            if self.grid[random_row][random_col] != 1:
                break
            if i > self.length*self.height*100:
                print(colored("Couldn't find random available position.", "red"))
                return None
                
        return((random_row, random_col))

    def get_start_goal(self, pHeuristic) -> Tuple[Tuple[int]]:
        i = 0

        # Initialization
        start = self.get_random_available_position()
        goal = self.get_random_available_position()
        n = Node(start[1], start[0], goal[1], goal[0], 0, False, False, None)
        h = n.calculate_heuristic()

        # Set goal value
        nTarget = Node(0, 0, self.length-1, self.height-1, 0, False, False, None)
        hTarget = nTarget.calculate_heuristic()

        while h < hTarget*pHeuristic:
            i += 1
            start = self.get_random_available_position()
            goal = self.get_random_available_position()
            n = Node(start[1], start[0], goal[1], goal[0], 0, False, False, None)
            h = n.calculate_heuristic()
            if i > 100*self.length*self.height:
                print(colored("Couldn't find start and goal respecting the constraints.", "red"))
                return None
        return start, goal
                
    def change_grid(self, position: Tuple[int], value: int):
        self.grid[position[1]][position[0]] = value
        
    def __generate_walls(self):
        for x in range(self.height):
            for y in range(self.length):
                if random.random() < self.p_walls:
                    self.grid[x][y] = 1
                if x == 0 or x == self.height - 1:
                    self.grid[x][y] = 1
                if y == 0 or y == self.length - 1:
                    self.grid[x][y] = 1

class Node:
    """
    >>> k = Node(0, 0, 4, 3, 0, None)
    >>> k.calculate_heuristic()
    5.0
    >>> n = Node(1, 4, 3, 4, 2, None)
    >>> n.calculate_heuristic()
    2.0
    >>> l = [k, n]
    >>> n == l[0]
    False
    >>> l.sort()
    >>> n == l[0]
    True
    """

    def __init__(self, pos_x: int, pos_y: int, goal_x: int, goal_y: int, g_cost: float, is_closed:bool, is_open:bool, parent):
        self.pos_x = pos_x
        self.pos_y = pos_y
        self.pos = (pos_y, pos_x)
        self.goal_x = goal_x
        self.goal_y = goal_y
        self.is_closed = is_closed
        self.is_open = is_open
        self.g_cost = g_cost
        self.parent = parent
        self.h_cost = self.calculate_heuristic()
        self.f_cost = self.g_cost + self.h_cost

    def calculate_heuristic(self) -> float:
        """
        Heuristic for the A*
        """
        dy = self.pos_x - self.goal_x
        dx = self.pos_y - self.goal_y
        if HEURISTIC == 1:
            return abs(dx) + abs(dy)
        else:
            return sqrt(dy ** 2 + dx ** 2)

    def print(self):
        print("Node : <Pos : ", self.pos, " >")

    def __lt__(self, other) -> bool:
        return self.f_cost <= other.f_cost


class AStar:
    """
    >>> wd = World(10, 10, 0.2)
    >>> astar = AStar((0, 0), (len(wd.grid) - 1, len(wd.grid[0]) - 1), wd)
    >>> (astar.start.pos_y + wd.delta[3][0], astar.start.pos_x + wd.delta[3][1])
    (0, 1)
    >>> [x.pos for x in astar.get_successors(astar.start)]
    [(1, 0), (0, 1)]
    >>> (astar.start.pos_y + wd.delta[2][0], astar.start.pos_x + wd.delta[2][1])
    (1, 0)
    >>> astar.retrace_path(astar.start)
    [(0, 0)]
    >>> astar.search()  # doctest: +NORMALIZE_WHITESPACE
    [(0, 0), (1, 0), (2, 0), (2, 1), (2, 2), (2, 3), (3, 3),
     (4, 3), (4, 4), (5, 4), (5, 5), (6, 5), (6, 6)]
    """

    def __init__(self, start: Tuple[int], goal: Tuple[int], world: World):
        self.start = Node(start[1], start[0], goal[1], goal[0], 0, False, True, None)
        self.target = Node(goal[1], goal[0], goal[1], goal[0], 99999, False, False, None)
        self.world = world

        self.nodes = dict()
        self.nodes[start] = self.start

        self.reached = False

    def search(self) -> List[Tuple[int]]:
        while self.count_open_nodes() > 0:
            # Open Nodes are sorted using __lt__
            current_key = min([n for n in self.nodes if self.nodes[n].is_open], key=(lambda k: self.nodes[k].f_cost))
            current_node = self.nodes[current_key]

            if current_node.pos == self.target.pos:
                print(colored("Found path.", "green"))
                return self.retrace_path(current_node)

            current_node.is_closed = True
            current_node.is_open = False

            successors = self.get_successors(current_node)

            for child_node in successors:
                if child_node.pos in self.nodes:
                    if self.nodes[child_node.pos].is_closed:
                        continue
                    if not self.nodes[child_node.pos].is_open:
                        self.nodes[child_node.pos] = child_node
                    else:
                        if child_node.g_cost < self.nodes[child_node.pos].g_cost:
                            self.nodes[child_node.pos] = child_node
                else:
                    self.nodes[child_node.pos] = child_node
        print(colored("Path not found", "red"))
        return [self.start.pos]

    def count_open_nodes(self) -> int:
        i = 0
        for _, val in self.nodes.items():
            if val.is_open:
                i += 1
                break
        return i

    def get_successors(self, parent: Node) -> List[Node]:
        """
        Returns a list of successors (both in the world and free spaces)
        """
        successors = []
        for action in self.world.delta:
            pos_x = parent.pos_x + action[0]
            pos_y = parent.pos_y + action[1]
            if not (0 <= pos_x < self.world.length and 0 <= pos_y < self.world.height):
                continue

            if self.world.grid[pos_y][pos_x] != 0:
                continue

            successors.append(
                Node(
                    pos_x,
                    pos_y,
                    self.target.pos_x,
                    self.target.pos_y,
                    parent.g_cost + action[2],
                    False,
                    True,
                    parent,
                )
            )
        return successors

    def retrace_path(self, node: Node) -> List[Tuple[int]]:
        """
        Retrace the path from parents to parents until start node
        """
        current_node = node
        path = []
        while current_node is not None:
            path.append((current_node.pos_y, current_node.pos_x))
            current_node = current_node.parent
        path.reverse()
        return path

if __name__ == "__main__":
    # all coordinates are given in format [y,x]
    # import doctest

    # doctest.testmod()

    w = World(70, 20, 0.2)
    w.plot_grid()
    
    # Generated
    start, goal = w.get_start_goal(0.8)

    # Hardcoded
    # start = (5, 5)
    # goal = (0, 9)
    
    print(start, goal)

    astar = AStar(start, goal, w)
    if TIME:
        st = time.process_time()
    path = astar.search()
    if TIME:
        print("Search() Execution Time : ", time.process_time() - st, "seconds")

    w.add_path(path)
    w.plot_grid()

File: Julia/astar.jl
module astar

DIAGONALS = true
HEURISTIC = 1

mutable struct Node
    pos::Int64
    goal_position::Int64
    fCost::Float64
    gCost::Float64
    hCost::Float64
    isClosed::Bool
    isOpen::Bool
    parent
end

mutable struct Path
    pos::Array{Int64}
    costs::Array{Float64}
end

mutable struct Environment
    length::Int64
    height::Int64
    wallPercentage::Float64
    grid::Array{Int8}
end

function get_actions(env)
    actions = [1 1; 
               env.length 1; 
              -1 1; 
              -env.length 1]
    
    if DIAGONALS
        d = [-env.length-1 sqrt(2); 
             -env.length+1 sqrt(2); 
              env.length-1 sqrt(2); 
              env.length+1 sqrt(2)]
        actions = [actions;d]
    end
    return actions
end

function print_color(color, string)
    if color == "red"
        print("\033[1;31m")
        print(string)
        print("\033[0m")
    elseif color == "blue"
        print("\033[1;34m")
        print(string)
        print("\033[0m")
    elseif color == "yellow"
        print("\033[1;33m")
        print(string)
        print("\033[0m")
    elseif color == "green"
        print("\033[1;32m")
        print(string)
        print("\033[0m")
    elseif color == "cyan"
        print("\033[1;36m")
        print(string)
        print("\033[0m")
    else
        print(string)
    end
end

function generate_environment(length, height, wallPercentage)
    gd = zeros(Int8, height, length)
    
    for (index, _) in enumerate(gd)
        if index % length == 0 || index <= length || index > length*height - length || (index - 1) % length == 0
            gd[index] = 1
            continue
        end
        if index//height % length == 0
            gd[index] = 1
            continue
        end
        if rand() < wallPercentage
            gd[index] = 1
        end

    end

    env = Environment(length, height, wallPercentage, gd)
    return env
end

function get_random_pos(env)
    size = env.length * env.height
    n = rand(1:size)
    while env.grid[n] == 1 
        n = rand(1:size)
    end
    return n
end

function get_start_goal(env, p_max_length)
    if p_max_length > 0.95
        p_max_length = 0.95
    end

    x = 0

    start = get_random_pos(env)
    goal = get_random_pos(env)
    size = env.length * env.height
    max_h = calculate_heuristic(env.length+2, size - env.length - 1, env)

    while calculate_heuristic(start, goal, env) <= max_h*p_max_length
        start = get_random_pos(env)
        goal = get_random_pos(env)
        x += 1
        if x > 1000
            print_color("red", "Could not find start and goal respecting constraints...\n")
            break
        end
    end

    return start, goal
end

function print_grid(env) 
    for (index, val) in enumerate(env.grid)
        if index % env.length != 0
            if val == 1
                print_color("red", "█")
            elseif val == 0
                print(" ")
            elseif val == 2
                print_color("blue", "¤")
            elseif val == 3
                print_color("green", "x")
            elseif val == 4
                print_color("cyan", "o")
            end
        else
            if val == 1
                print_color("red", "█")
            elseif val == 0
                print(" ")
            elseif val == 2
                print_color("blue", "¤")
            elseif val == 3
                print_color("green", "x")
            elseif val == 4
                print_color("cyan", "o")
            end
            print("\n")
        end
    end
end

function add_elem_to_env(env, pos, val)
    env.grid[pos] = val
    return env
end

function add_path(env, path)
    for (index, val) in enumerate(path.pos)
        if val == path.pos[1] || val == last(path.pos)
            continue
        end
        env = add_elem_to_env(env, val, 2)
    end

    return env
end

function calculate_heuristic(current_pos, goal_pos, env)
    row_current = convert(Int64, floor(current_pos / env.length)) + 1
    col_current = current_pos % env.length
    row_target = convert(Int64, floor(goal_pos / env.length)) + 1
    col_target = goal_pos % env.length

    dy = abs(row_target - row_current)
    dx = abs(col_target - col_current)

    if HEURISTIC == 0
        return(sqrt(dy^2 + dx^2))
    elseif HEURISTIC == 1
        return(dx + dy)
    else
        print_color("red", "Reached unreachable statement.")
        exit()
    end
end

function find_lowest_cost_node(nodes) 
    x = 0
    currentNode = nothing
    for (_, val) in nodes
        if val.isOpen
            x += 1
            currentNode = val
            break
        end
    end

    if !(x > 0)
        return 0, nothing
    end

    for (_, val) in nodes
        if val.isOpen && val != currentNode 
            x += 1
            if val.fCost < currentNode.fCost
                currentNode = val
            end
        end
    end

    return x, currentNode
end

function get_neighbours(current_node, goal_pos, env)
    c_pos = current_node.pos
    size = env.length * env.height

    children = []
    
    if c_pos < 0 || c_pos >= env.length*env.height || env.grid[c_pos] == 1
        print_color("red", "Tried to get neighbours from unreachable tile, exiting...")
        exit()
    end

    for row in eachrow(get_actions(env))
        if !(c_pos + row[1] > 0 && c_pos + row[1] <= size)
            continue
        end

        if env.grid[Int(c_pos + row[1])] == 1
            continue
        end

        child = node_constructor(Int(c_pos + row[1]), 
                                 goal_pos, 
                                 current_node.gCost + row[2], 
                                 false, false, current_node, env)
        
        children = push!(children, child)
    end
    
    return children
end

function node_constructor(position, goal_position, gcost, isclosed, isopen, parent, env)
    h = calculate_heuristic(position, goal_position, env)
    return(Node(position, goal_position, gcost + h, gcost, h, isclosed, isopen, parent))
end

function retrieve_path(last_node)
    current_node = last_node
    path = Path([], [])
    while current_node !== nothing
        path.pos = push!(path.pos, current_node.pos)
        path.costs = push!(path.costs, current_node.gCost)
        current_node = current_node.parent
    end
    return path
end

function search(start, goal, env)
    nodes = Dict{Int64, Node}()
    start_node = node_constructor(start, goal, 0, false, true, nothing, env)
    nodes[start] = start_node

    while true
        open_nodes_cnt, currentNode = find_lowest_cost_node(nodes)
        
        if open_nodes_cnt < 1
            print_color("red", "The graph has no open nodes left...\n")
            break
        end

        if currentNode.pos == goal
            print_color("yellow", "Path found!\n")
            return(retrieve_path(currentNode))
        end
        
        nodes[currentNode.pos].isOpen = false
        nodes[currentNode.pos].isClosed = true

        children = get_neighbours(currentNode, goal, env)

        for (_, child) in enumerate(children)
            if haskey(nodes, child.pos)
                if nodes[child.pos].isClosed
                    continue
                end

                if !nodes[child.pos].isOpen
                    child.isOpen = true
                    nodes[child.pos] = child
                else
                    if child.gCost < nodes[child.pos].gCost
                        child.isOpen = true
                        nodes[child.pos] = child
                    end
                end
            else
                child.isOpen = true
                nodes[child.pos] = child
            end
        end
    end
end

function main()
    println(PROGRAM_FILE, " starting...")
    length = 50
    height = 20
    wallP = .25
    env = generate_environment(length, height, wallP)

    start, goal = get_start_goal(env, 0.5)

    env = add_elem_to_env(env, start, 3)
    env = add_elem_to_env(env, goal, 3)

    print_grid(env)

    path = search(start, goal, env)

    if path !== nothing 
        env = add_path(env, path)
        print_grid(env)
    end
end

main()

end #astar